cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go

```diff
@@ -26,73 +26,179 @@ import (
+	chartList, err := resourceIfc.List(ctx, metav1.ListOptions{})
 	if err != nil {
-		return false, err
-	} else if !found {
-		return false, nil
+		return nil, err
 	}
-	if generation != observedGeneration {
-		return false, nil
+	return chartList, nil
+}
+
+// returns the url from which chart .tgz can be downloaded
+// here chartVersion string, if specified at all, should be specific, like "14.4.0",
+// not an expression like ">14 <15"
+func (s *Server) getChartTarball(ctx context.Context, repoName string, chartName string, namespace string, chartVersion string) (url string, err error, cleanUp func()) {
+	// see if we the chart already exists
+	// TODO (gfichtenholt):
+	// see https://github.com/kubeapps/kubeapps/pull/2915
+	// for context. It'd be better if we could filter on server-side. The problem is the set of supported
+	// fields in FieldSelector is very small. things like "spec.chart" or "status.artifact.revision" are
+	// certainly not supported.
+	// see
+	//  - kubernetes/client-go#713 and
+	//  - https://github.com/flant/shell-operator/blob/8fa3c3b8cfeb1ddb37b070b7a871561fdffe788b///HOOKS.md#fieldselector and
+	//  - https://github.com/kubernetes/kubernetes/issues/53459
+	chartList, err := s.listChartsInCluster(ctx, namespace)
+	if err != nil {
+		return "", err, nil
 	}
 
-	conditions, found, err := unstructured.NestedSlice(unstructuredChart.Object, "status", "conditions")
+	url, err = findUrlForChartInList(chartList, repoName, chartName, chartVersion)
 	if err != nil {
-		return false, err
-	} else if !found {
-		return false, nil
-	}
-
-	// check if ready=True
-	for _, conditionUnstructured := range conditions {
-		if conditionAsMap, ok := conditionUnstructured.(map[string]interface{}); ok {
-			if typeString, ok := conditionAsMap["type"]; ok && typeString == "Ready" {
-				if statusString, ok := conditionAsMap["status"]; ok {
-					if statusString == "True" {
-						if reasonString, ok := conditionAsMap["reason"]; !ok || reasonString != "ChartPullSucceeded" {
-							// should not happen
-							log.Infof("unexpected status of HelmChart: %v", *unstructuredChart)
-						}
-						return true, nil
-					} else if statusString == "False" {
-						var msg string
-						if msg, ok = conditionAsMap["message"].(string); !ok {
-							msg = fmt.Sprintf("No message available in condition: %v", conditionAsMap)
-						}
-						// chart pull is done and it's a failure
-						return true, status.Errorf(codes.Internal, msg)
-					}
+		return "", err, nil
+	} else if url != "" {
+		return url, nil, nil
+	}
+
+	// did not find the chart, need to create
+	// see https://fluxcd.io/docs/components/source/helmcharts/
+	// TODO (gfichtenholt)
+	// 1. HelmChart object needs to be co-located in the same namespace as the HelmRepository it is referencing.
+	// 2. flux impersonates a "super" user when doing this (see fluxv2 plug-in specific notes at the end of
+	//	design doc). We should probably be doing simething similar to avoid RBAC-related problems
```

Comment:
FWIW, I don't think we'll ever include a `cluster-admin` service account with Kubeapps or a plugin - that's just too dangerous, as you can see from the [flux proposal](https://github.com/fluxcd/flux2/pull/582/files#diff-e377d0286cdc254b5541414d2db6f291be9d92a669ccb2d4e7fb14a8645dfa57) (that hasn't been accepted yet) which you linked to in the doc, about how they're trying to get out of that situation.

Helm itself used to do something similar, having `tiller` running in the cluster with crazy privileges that could be used to do exactly what they describe in the flux proposal... I'm quite surprised that they have `cluster-admin` as the default for the helm controller TBH, given the history.
Anyway, fine as is, as long as we're not wanting to deploy a kubeapps component with super privs.
So back to here: The user carrying out the request will need to have RBAC for whatever is done - currently it looks like you're creating the flux HelmChart CR explicitly. If that is necessary (see previous comment above as I'm not sure why it is if the HelmRelease already exists) we can document that as a requirement.

Reply:
+1. If you'd like me to word the comment a certain way, just say so.

Reply:
I think the comment is fine as a note - word it however you like :) I just don't think option 2 is something we should do (even flux seems to be backing away from it).